types/rlp_unmarshal.go

```diff
@@ -376,44 +381,64 @@ func (t *Transaction) unmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) erro
 		return err
 	}
 
-	if len(elems) < 9 {
-		return fmt.Errorf("incorrect number of elements to decode transaction, expected 9 but found %d", len(elems))
+	getElem := func() *fastrlp.Value {
+		val := elems[0]
+		elems = elems[1:]
+
+		return val
+	}
+
+	var num int
+
+	switch t.Type {
+	case LegacyTx:
+		num = 9
+	case DynamicFeeTx:
+		num = 10
+	case StateTx:
+		num = 10
+	default:
+		return fmt.Errorf("transaction type %d not found", t.Type)
+	}
+
+	if numElems := len(elems); numElems != num {
+		return fmt.Errorf("incorrect number of transaction elements, expected %d but found %d", num, numElems)
 	}
 
 	p.Hash(t.Hash[:0], v)
 
 	// nonce
-	if t.Nonce, err = elems[0].GetUint64(); err != nil {
+	if t.Nonce, err = getElem().GetUint64(); err != nil {
```

Comment:
Am I missing something? is the `getElem` dynamic? how it knows the element index?

Reply:
Kind of. Dynamic base fee tx has two more fields instead of gas price. Basically, take a look at the code above to find an elements number depending on tx type.

Reply:
I see, LGTM